src/colmap/estimators/cost_functions.h

```diff
@@ -301,6 +302,111 @@ class SampsonErrorCostFunction {
+    const Eigen::Quaternion<T> mes_from_est_q =
+        cam_from_world_.rotation.cast<T>() *
+        EigenQuaternionMap<T>(cam_from_world_q).inverse();
+    EigenQuaternionToAngleAxis(mes_from_est_q.coeffs().data(), residuals_ptr);
+
+    Eigen::Map<Eigen::Matrix<T, 3, 1>> mes_from_est_t(residuals_ptr + 3);
+    mes_from_est_t = cam_from_world_.translation.cast<T>() -
+                     mes_from_est_q * EigenVector3Map<T>(cam_from_world_t);
+
+    Eigen::Map<Eigen::Matrix<T, 6, 1>> residuals(residuals_ptr);
+    residuals.applyOnTheLeft(sqrt_information_.template cast<T>());
+    return true;
+  }
+
+ private:
+  const Rigid3d& cam_from_world_;
+  const EigenMatrix6d sqrt_information_;
+};
+
+// 6-DoF error between two absolute poses based on a measurement that is their
+// relative pose, with identical scale for the translation. The covariance is
+// defined in the reference frame of the camera j.
+// Its first and last three components correspond to the rotation and
+// translation errors, respectively.
+struct MetricRelativePoseErrorCostFunction {
+ public:
+  explicit MetricRelativePoseErrorCostFunction(const Rigid3d& j_from_i,
+                                               const EigenMatrix6d& covariance)
+      : j_from_i_(j_from_i),
+        sqrt_information_(covariance.inverse().llt().matrixL()) {}
+
+  static ceres::CostFunction* Create(const Rigid3d& j_from_i,
+                                     const EigenMatrix6d& covariance) {
+    return (new ceres::AutoDiffCostFunction<MetricRelativePoseErrorCostFunction,
+                                            6,
+                                            4,
+                                            3,
+                                            4,
+                                            3>(
+        new MetricRelativePoseErrorCostFunction(j_from_i, covariance)));
+  }
+
+  template <typename T>
+  bool operator()(const T* const i_from_world_q,
+                  const T* const i_from_world_t,
+                  const T* const j_from_world_q,
+                  const T* const j_from_world_t,
+                  T* residuals_ptr) const {
+    const Eigen::Quaternion<T> i_from_j_q =
+        EigenQuaternionMap<T>(i_from_world_q) *
+        EigenQuaternionMap<T>(j_from_world_q).inverse();
+    Eigen::Matrix<T, 3, 1> i_from_j_t =
+        EigenVector3Map<T>(i_from_world_t) -
+        i_from_j_q * EigenVector3Map<T>(j_from_world_t);
+
+    const Eigen::Quaternion<T> mes_from_est_q =
+        j_from_i_.rotation.cast<T>() * i_from_j_q;
+    EigenQuaternionToAngleAxis(mes_from_est_q.coeffs().data(), residuals_ptr);
+
+    Eigen::Map<Eigen::Matrix<T, 3, 1>> mes_from_est_t(residuals_ptr + 3);
+    mes_from_est_t = j_from_i_.translation.cast<T>() +
+                     j_from_i_.rotation.cast<T>() * i_from_j_t;
+
+    Eigen::Map<Eigen::Matrix<T, 6, 1>> residuals(residuals_ptr);
+    residuals.applyOnTheLeft(sqrt_information_.template cast<T>());
```

Comment:
We need to be very careful here about the definition of covariance here. Is it defined in frame i or frame j. It will be good to explicitly express this in the variable name of the covariance and sqrt_information variables (i.e., covariance_in_j)

Reply:
The docstring mentions that the covariance is defined in the reference frame of the camera $j$. I updated the name of the covariance accordingly.